### ISSUE TITLE
[Feature Request]: 

### ISSUE THREAD

**Issue by YuriL180821** [electron/electron#30589]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

Support Alert type notifications on MacOS by the Electron with usage of NSNotificationCenter requires App-XPC Service scheme(similar like in Chromium). XPC Service requires to be integrated into App/Contents. Current Electron ninja build scripts do not have any XPC service section. Need to have support for XPC service skeleton into Electron project.

### Proposed Solution

Do not have solution for this, expecting explanations or concrete sample for the XPC service integration into ninja build scripts from electron community

### Alternatives Considered

Was made few tries to adopt build script existing alert_notification_xpc_service from Chromium but with no success (XPC service was not present into Electron app.package)
Build script was taken from 
%cd%/chrome/browser/ui/cocoa/notifications/BUILD.gn

### Additional Information

Reuse of existing XPC service from chromium will require support for all Chromium eco-system such as
chrome\browser\notifications
chrome\browser\ui\cocoa\notifications
Chromium Notification concept is differ that current Electron's one. And even in case of reuse Chromium XPC Alert Service required to be integrated into Electron, that means changing existing Electron build scripts too. This variant looks more complicated than variant of creation Electron specific alert Service.

**Comment by AditiThirdEye** [comment]
Can this task be assigned to me?

**Comment by YuriL180821** [comment]
Hello @AditiThirdEye,  

I'm not sure that question was addressed to me, but in case if You have expertise with ninja build scripts You are welcome. Few words again about the problem: **implementation for the Alert XPC Service is not necessary, I already have it**. Just need to have explanations or gn/gni files set which allow to integrate any XPC service into electron app(XPC can be even simple example from MacOS dev manual https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingXPCServices.html)

**Comment by AditiThirdEye** [comment]
Hello @YuriL180821,
I don't have expertise in ninja build scripts, I was just intrigued when an opportunity came to add a feature in electron hence wanted to give it a try. 
Your example on MacOS dev manual, I'll look into it. 

Thank and regards,
Aditi Pandey

**Comment by nornagon** [comment]
Can you say a bit more about what the end result is that you want to achieve?

**Comment by AditiThirdEye** [comment]
Yes so in our college there is a task to add a feature or extension in a repo with good number of stars so wanted to work on this. Didn't had a particular result to be achieved in the end of it.

Sorry for the inconveniences.

**Comment by nornagon** [comment]
@AditiThirdEye sorry, I was asking the initial commenter @YuriL180821 to provide more details on their feature request.

**Comment by YuriL180821** [comment]
Hello @nornagon ,

Chrome which uses same core contains embedded Alert XPC service 
**Chromium.app/Contents/Frameworks/Chromium/Framework.framework/Versions/A/XPCServices/**

My project needs demands similar integration XPC service but into Electron app. Currently I've got stuck into modification ninja build scripts which allows inclusion my own XPC Alert service(I'm not considering reuse of existing alert service from chromium because Electron has different concept for notification and adaptation of existing service will cost much greater that creation of the new one) .

Detalization:
As was mentioned above Chrome has Alert service into it's internals. Build script for this service is located here
**%cd%/chrome/browser/ui/cocoa/notifications/BUILD.gn**

Build script contains few sections such as

`import("//build/config/mac/rules.gni")
import("//build/util/branding.gni")

mac_app_bundle("alert_notification_xpc_service") {
  output_name = "AlertNotificationService"
  package_type = "xpc"

  info_plist = "xpc-Info.plist"
  extra_substitutions = [ "CHROME_BUNDLE_ID=$chrome_mac_bundle_id" ]`
  
  ...
  `ldflags = []
  if (is_component_build) {
    ldflags += [
      # The XPC service is in Chromium.app/Contents/Frameworks/Chromium \
      # Framework.framework/Versions/A/XPCServices/ \
      # AlertNotificationService.xpc/Contents/MacOS/
      # so set rpath up to the base.
      "-rpath",
      "@loader_path/../../../../../../../../../../",
    ]
  }
}`

e.t.c

I've tried to modify this script and include into electron project coupled with my new XPC Alert service (i.e. I've added XPC service code and **build.gn** script) but after compilation Electron app did not contain my XPC service inside. So I do not understand what I need to do for integration my XPC service inside of Electron.

**Comment by zcbenz** [comment]
It seems that Chrome browser no longer uses any xpc service anymore?
https://source.chromium.org/search?q=xpc_service&ss=chromium

Also I think to show a notification in alert type the official way is to set `NSUserNotificationAlertStyle` in Info.plist?
https://stackoverflow.com/questions/11831395/style-of-nsusernotification

**Comment by sadpixel** [comment]
Y'all, this already exists and is supported. Just use the HTML5 Notification support baked into Electron: https://www.electronjs.org/docs/tutorial/notifications

On macOS, you'll get the native notification style, grouped in Notification Center under your app's own section. 😘👌 You can even trigger them from any embedded page using simple JS tunneled through ipcMain.on()

**Comment by YuriL180821** [comment]
Hello all!

@zcbenz zcbenz
Regarding to this **_NSUserNotificationAlertStyle in Info.plist?_** it will change behavior for banners, i.e banners will stay on Desktop as Alerts. That is why XPC service is necessary : to separate  processing for Banners and Alerts

XPC service is necessary thing to support Alerts on old MacOS platforms (I mean where UNNotificcations are not available), modern 11-th API allows to show buttons/replies even with Notifications scheme Banner. 

**Does anyone can help me with initial question ? It's still open.**

**Comment by sadpixel** [comment]
> @sadpixel Electron does not have support for persistent notifications, yet...

What do you mean by persistent notifications? A notification that appears when the app isn't running, like a notification that you have a new iMessage or email when Messages.app or Mail.app aren't open?

If that's what you meant, let me know, as there *is* a simply workaround to accomplish just that...

**Comment by YuriL180821** [comment]
Hello @sadpixel !

I mean that Electron can't display notification with buttons/reply fileds e.t.c that means persistent.
My question was reharding to XC service integration(not creation) without any workarounds(I've tested much more that was shared into that topic).

Regards.

**Comment by nornagon** [comment]
@MarshallOfSound I think this is maybe covered by the eventual refactor from NSNotification -> UNUserNotificationCenter, right?
